### PR TITLE
feature: apply filters on response that want to be processed

### PR DESCRIPTION
**Pull request by szuecs**
feature: apply filters on response that want to be processed
fix: admissionControl 499 cases should be counted

- [x] tests
- [x] docs

### CODE REVIEW

File: proxy/proxy.go [hunk 1]
```diff
@@ -1272,6 +1279,9 @@ func (p *Proxy) errorResponse(ctx *context, err error) {
 	if ctx.response.StatusCode == 499 {
 		msgPrefix = "client canceled"
 		logFunc = p.log.Infof
+		if p.accessLogDisabled {
+			logFunc = p.log.Debugf
+		}
```
Comment (AlexanderYastrebov) [inline proxy/proxy.go:1284]: Why this change in this scope?

Comment (szuecs) [reply proxy/proxy.go:1284]: Because I think we did not caught it the last time and I saw it during tests.

File: filters/filters.go [hunk 2]
```diff
@@ -151,6 +151,15 @@ type Filter interface {
...
 }
 
+// FilterCloser are Filters that need to cleanup resources after
+// filter teminiation. For example Filters, that create a goroutine
+// for some reason need to cleanup their goroutine or they would leak
+// goroutines.
+type FilterCloser interface {
+	Filter
+	Close()
+}
```
Comment (AlexanderYastrebov) [inline filters/filters.go:161]: We implement goroutine cleanup via admissionControlPost

Maybe in your test you could utilize it and e.g. update routing table to an empty value just before test ends such that post-processor closes all instances.

Alternatively we may call cleanup manually on the post-postprocessor like it is done here https://github.com/zalando/skipper/blob/e57538ee7028f79baa0a285db268c4e7c3cf346e/filters/auth/main_test.go#L11-L17

Comment (AlexanderYastrebov) [reply filters/filters.go:161]: This is related to the long known but hard problem of filter resource cleanup https://github.com/zalando/skipper/issues/202

Comment (szuecs) [reply filters/filters.go:161]: How can you access the PostProcessor from the test?

File: proxy/failing_filter_test.go [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+	}))
+	defer backend.Close()
+
+	var routes = fmt.Sprintf(`
+                main: * -> admissionControl("mygroup", "active", "1s", 5, 0, 0.99, 0.9, 1.0) -> "%s";
+        `, backend.URL)
+	r := eskip.MustParse(routes)
+
+	fr := make(filters.Registry)
+	fr.Register(shedder.NewAdmissionControl(shedder.Options{}))
```
Comment (AlexanderYastrebov) [inline proxy/failing_filter_test.go:51]: I think we need to create and configure postProcessor for it.

Comment (szuecs) [reply proxy/failing_filter_test.go:51]: Yes true that might help

Comment (szuecs) [reply proxy/failing_filter_test.go:51]: I tested it and it does not help. PostProcessor is only working on routing updates.

File: filters/filters.go (hunk 2 above)
Comment (szuecs) [reply filters/filters.go:161]: I found a way to fix all the things. It did not require a TestMain hack.

File: proxy/proxy.go [hunk 4]
```diff
@@ -1551,3 +1561,36 @@ func (p *Proxy) makeErrorResponse(ctx *context, perr *proxyError) {
...
+
+func (p *Proxy) applyFiltersOnError(ctx *context, filters []*routing.RouteFilter) {
+	filtersStart := time.Now()
+	filterTracing := p.tracing.startFilterTracing("response_filters", ctx)
+	defer filterTracing.finish()
+
+	for i := len(filters) - 1; i >= 0; i-- {
+		fi := filters[i]
+
+		if _, ok := fi.Filter.(errorHandlerFilter); !ok {
```
Comment (AlexanderYastrebov) [inline proxy/proxy.go:1580]: I think here we want
```go
if ehf, ok := fi.Filter.(errorHandlerFilter); !ok || !ehf.HandleErrorResponse() {
  continue
}
```

File: routing/datasource.go [hunk 5]
```diff
@@ -585,6 +586,23 @@ func receiveRouteMatcher(o Options, out chan<- *routeTable, quit <-chan struct{}
...
+			}
+			for _, r := range routes {
+				for _, f := range r.Filters {
+					fc, ok := f.Filter.(filters.FilterCloser)
+					if ok {
+						fc.Close()
+					}
+				}
+			}
+
```
Comment (AlexanderYastrebov) [inline routing/datasource.go:605]: I guess we can just keep a pointer to the routes:
```diff
diff --git a/routing/datasource.go b/routing/datasource.go
index 9b679df4..c2ab4ace 100644
--- a/routing/datasource.go
+++ b/routing/datasource.go
@@ -520,6 +520,7 @@ func processRouteDefs(o Options, fr filters.Registry, defs []*eskip.Route) (rout
 
 type routeTable struct {
        m             *matcher
+       routes        []*Route
        validRoutes   []*eskip.Route
        invalidRoutes []*eskip.Route
        created       time.Time
@@ -574,6 +575,7 @@ func receiveRouteMatcher(o Options, out chan<- *routeTable, quit <-chan struct{}
 
                        rt = &routeTable{
                                m:             m,
+                               routes:        routes,
                                validRoutes:   validRoutes,
                                invalidRoutes: invalidRoutes,
                                created:       time.Now().UTC(),
@@ -585,6 +587,13 @@ func receiveRouteMatcher(o Options, out chan<- *routeTable, quit <-chan struct{}
                        updatesRelay = updates
                        outRelay = nil
                case <-quit:
+                       for _, route := range rt.routes {
+                               for _, f := range route.Filters {
+                                       if fc, ok := f.Filter.(filters.FilterCloser); ok {
+                                               fc.Close()
+                                       }
+                               }
+                       }
                        return
                }
        }
```

Comment (AlexanderYastrebov) [reply routing/datasource.go:605]: I think we can do that in the startReceivingUpdates

```go
--- a/routing/routing.go
+++ b/routing/routing.go
@@ -371,6 +371,10 @@ func (r *Routing) startReceivingUpdates(o Options) {
                                }
                                r.log.Info("route settings applied")
                        case <-r.quit:
+                               rt := r.routeTable.Load().(*routeTable)
+                               if rt != nil {
+                                       // ...
+                               }
                                return
                        }
                }
```

File: proxy/proxy.go (hunk 4 above)
Comment (szuecs) [reply proxy/proxy.go:1580]: done

File: routing/datasource.go (hunk 5 above)
Comment (szuecs) [reply routing/datasource.go:605]: The problem with routeTable is that we do not have all routes here, only access to leafRoutes as far as I tried last week.

Comment (szuecs) [reply routing/datasource.go:605]: I guess you suggest to add the routes to the routeTable to store the link to all the routes, which would be good enough to fix the problem, yes.

Comment (AlexanderYastrebov) [reply routing/datasource.go:605]: yeah, keep the link and close filters from startReceivingUpdates

File: filters/filters.go [hunk 6]
```diff
@@ -151,6 +151,15 @@ type Filter interface {
 	Response(FilterContext)
 }
 
+// FilterCloser are Filters that need to cleanup resources after
+// filter teminiation. For example Filters, that create a goroutine
+// for some reason need to cleanup their goroutine or they would leak
+// goroutines.
+type FilterCloser interface {
+	Filter
+	Close()
```
Comment (AlexanderYastrebov) [inline filters/filters.go:160]: Maybe it should be `Close() error` (i.e. io.Closer) before its too late - we can always ignore error but it would be hard to introduce it later.

File: routing/datasource.go (hunk 5 above)
Comment (szuecs) [reply routing/datasource.go]: Thanks for the input I put it into Routing.Close, because there it makes most sense and then we do not have the problem of nil ptr.
Let's see what the tests say.

File: routing/routing.go [hunk 7]
```diff
@@ -422,6 +422,11 @@ func (r *Routing) Get() *RouteLookup {
 
 // Close closes routing, stops receiving routes.
 func (r *Routing) Close() {
+	var rt *routeTable
+	rt, ok := r.routeTable.Load().(*routeTable)
+	if ok {
+		rt.Close()
+	}
```
Comment (AlexanderYastrebov) [inline routing/routing.go:429]: I think it is better to put it into startReceivingUpdates into `case <-r.quit:` - this way we know there will be no updates anymore

File: routing/datasource.go [hunk 8]
```diff
@@ -520,11 +521,27 @@ func processRouteDefs(o Options, fr filters.Registry, defs []*eskip.Route) (rout
...
+	once          sync.Once
+	routes        []*Route // only used for closing
 	validRoutes   []*eskip.Route
 	invalidRoutes []*eskip.Route
 	created       time.Time
 }
 
+// Close routeTable will cleanup all underlying resources, that could
+// leak goroutines.
+func (rt *routeTable) Close() {
```
[the remark this answers is not in the document]
Comment (szuecs) [reply routing/datasource.go:533]: yes but it looks so ugly in case you have syntax highlighting, because `close` is an internal func like println. I agree will make it lower case, but I wait to finish the tests

File: routing/routing.go (hunk 7 above)
Comment (szuecs) [reply routing/routing.go:429]: It's not possible because routeTable is nil, just check the 2-3 lines above the `<-quit` case, there we set `rt = nil`

Comment (AlexanderYastrebov) [reply routing/routing.go]: I mean in the startReceivingUpdates, not in the receiveRouteMatcher

Comment (szuecs) [reply routing/routing.go]: changed, let's see what tests are saying. One problem could be that it's async closing by another goroutine and goroutine leak detection might fail

File: docs/tutorials/development.md [hunk 9]
```diff
@@ -137,6 +137,86 @@ func (f *myFilter) Response(ctx filters.FilterContext) {
 
 Find a detailed example at [how to develop a filter](../reference/development.md#how-to-develop-a-filter).
 
+### Filters with cleanup
```
Comment (AlexanderYastrebov) [inline docs/tutorials/development.md:140]: I am not sure we should document it here at this time as this might create confusion for the users.
This change only closes filters on the routing shutdown, i.e. once in the Skipper lifetime while filters could be recreated on each route update. Users may get confused and think Close will be called for each their filter instance that is not used anymore.

The reason it works for admission control is because there is a postprocesor that closes unused filters (which technically is still incorrect as there might be in-flight requests during route update)

I.e. it does not solve https://github.com/zalando/skipper/issues/202 but only makes a small step towards.

File: filters/filters.go (hunk 6 above)
Comment (AlexanderYastrebov) [reply filters/filters.go:160]: With returned error you can decide to ignore it or handle. Without error there is no way to handle an error if any.

**Comment by AlexanderYastrebov** [comment]
:+1:

File: filters/filters.go (hunk 6 above)
Comment (szuecs) [reply filters/filters.go]: fixed

File: docs/tutorials/development.md (hunk 9 above)
Comment (szuecs) [reply docs/tutorials/development.md:140]: good point about the incorrect closing. Maybe we should introduce a closing registry which will wait for timeouts like ResponseHeaderTimeout long before it closes the filters and we move in the PostProcessor the "Close" call to something like closingRegistry.Add(filterInstance). wdyt?

File: docs/tutorials/development.md [hunk 10]
```diff
@@ -137,6 +137,86 @@ func (f *myFilter) Response(ctx filters.FilterContext) {
...
+
+func (f *myFilter) Request(ctx filters.FilterContext) {
+     // change data in ctx.Request() for example
+}
+
+func (f *myFilter) Response(ctx filters.FilterContext) {
+     // change data in ctx.Response() for example
+}
+
+func (f *myFilter) Close() {
```
Comment (AlexanderYastrebov) [inline docs/tutorials/development.md:169]: Should return error

File: filters/filters.go (hunk 6 above)
Comment (szuecs) [reply filters/filters.go]: fixed

File: docs/tutorials/development.md (hunk 9 above)
Comment (szuecs) [reply docs/tutorials/development.md:140]: a sync.Atomic as gauge per route?

File: docs/tutorials/development.md (hunk 10 above)
Comment (szuecs) [reply docs/tutorials/development.md:169]: fixed

File: filters/filters.go [hunk 11]
```diff
@@ -151,6 +151,15 @@ type Filter interface {
 	Response(FilterContext)
 }
 
+// FilterCloser are Filters that need to cleanup resources after
+// filter teminiation. For example Filters, that create a goroutine
```
[the remark this answers is not in the document]
Comment (szuecs) [reply filters/filters.go:155]: fixed

**Comment by szuecs** [comment]
:+1:

**Comment by AlexanderYastrebov** [comment]
:+1:
